### ISSUE TITLE
Ecto.Query.from/2 in pipeline report refactor warning

### ISSUE THREAD

**Issue by CaptChrisD** [rrrene/credo#280]
### Environment

* Credo version (`mix credo -v`): 0.5.3
* Erlang/Elixir version (`elixir -v`): 1.3.4
* Operating system: macOS 10.12.1

### What were you trying to do?

With a `Ecto.Query.from/2` call like

```elixir
    (from comment in Comments,
      where: comment.id == ^id)
    |> Repo.one
```

### Expected outcome

No warning should be reported

### Actual outcome

Credo reports a Refactor opportunity of "Pipe chain should start with a raw value"

Unless I am missing something, it doesn't seem like this should be a warning. Yes you could assign the return of from to a variable and then pipeline, but in this case that would defeat the purpose of the pipeline here

**Comment by devonestes** [comment]
If I'm not mistaken, I think the check in this case is behaving correctly. The way the check is set up is that pipe chains have to begin with some sort of single value, like below:

```elixir
str
  |> String.downcase
  |> String.reverse
```

There's actually another check that's coming soon that will also report a Refactor opportunity on that code for using a pipe only once. See José's issue on that here: https://github.com/rrrene/credo/issues/221

I think it doesn't really defeat the purpose of the pipe to assign something to a variable - there's a real readability benefit there that I think can be quite helpful. But, given the single pipe check, I think the below might be a good way to write that code (excuse me if this isn't 100% valid - I haven't used much of Ecto yet):

```elixir
Repo.one(from comment in Comments, where: comment.id == ^id)
```

Personally I think that's a little more readable, but that's just me 😉  Also, these aren't really hard and fast rules - more just guidelines that are meant to be customized as needed - so if you disagree with a given check you can always disable it or lower its priority significantly.

**Comment by rrrene** [comment]
@CaptChrisD you are right. we addressed this on `master` but I did not manage to make a proper new release yet.

You could do me a favor: Set the dependency to `{:credo, github: "rrrene/credo"}` and see if the "Refactoring opportunity" disappears for the bleeding edge version of Credo.

@devonestes we added the "parentheses exception" for cases like these

    block =
      (options[:do] || contents[:do])
      |> prewalk_block(name)
      |> handler_builder(type)

**Comment by CaptChrisD** [comment]
@rrrene I have tried the bleeding edge version and have in my lock file `"credo": {:git, "https://github.com/rrrene/credo.git", "631a5fa4e98071892d0646ab376c7cf77848bbd9", []}`
I still get the refactor opportunities.

@devonestes In the example I gave the from is quite small and inline isn't so bad, but on complicated queries the `from` command may be quite large and not clear when inlined

**Comment by jmitchell** [comment]
I bumped into this as well and made the following minimal test case:

```
  def dummy(x) do
    (x == 5) |> IO.put
  end
```

The undesirable result:

```
┃ [F] → Pipe chain should start with a raw value.
┃       lib/credo.ex:3 (CredoTest.dummy)
┃
┃           (x == 5) |> IO.put
```

This reproduces with the latest `master` revision. From my `mix.lock`.

```
"credo": {:git, "https://github.com/rrrene/credo.git", "917ea4c416577754d070b0356208d7fa7aa2c5ae", []}
```

**Comment by jmitchell** [comment]
On further reflection, the problem generalizes to binary infix operators or DSL expressions where it's awkward to extract the first argument.

My minimal case, for instance, could be "fixed" by doing this, but it's awkward because `(x == 5)` is much more readable.

```
x |> Kernel.==(5) |> IO.put
```

**Comment by jmitchell** [comment]
Here's another example of the general problem, this time with a comprehension. The use case is collecting the special 9 cell regions on a Sudoku puzzle.

```elixir
def regions do
  (for i <- 0..8, do: [row(i), column(i), sector(i)])
  |> Stream.concat()
end
```

The way to work around the warning is to make a dummy variable, but I think it adds unnecessary noise.

```elixir
def regions do
  rcs = for i <- 0..8, do: [row(i), column(i), sector(i)]

  rcs
  |> Stream.concat()
end
```

**Comment by dylan-chong** [comment]
I am having another problem with the same pipeline error.

```
      Stream.repeatedly(fn -> char end)
      |> Enum.take(n)
      |> to_string()
```

Why is it showing up with an error here?

**Comment by rrrene** [comment]
@dylan-chong You have to explicitly exclude `Stream.repeatedly` using the param `excluded_functions:`.

Example:  https://github.com/rrrene/credo/blob/cd946b8ef07fa3080b16d3f0ae13c77dc6dddb70/test/credo/check/refactor/pipe_chain_start_test.exs#L143

**Comment by rrrene** [comment]
@jmitchell This is fixed now. I can not see when it was fixed, but we added for [this exact scenario](https://github.com/rrrene/credo/blob/cd946b8ef07fa3080b16d3f0ae13c77dc6dddb70/test/credo/check/refactor/pipe_chain_start_test.exs#L113) so I will mark this as done.

Thx for reporting your issues! :+1:

**Comment by wafcio** [comment]
@rrrene I have the same problem with version 0.8.0 and with master branch

**Comment by rrrene** [comment]
@wafcio Please see my last answer in this thread. You can exclude functions from the check like that in `.credo.exs` (or exclude the check completely).

**Comment by dylan-chong** [comment]
I don't believe manually excluding functions one by one is a very good solution

Can't we change the defaults by type? For example,  ignore situations where the parameter is a function or a keyword list,

**Comment by rrrene** [comment]
@dylan-chong I see your point. Could write that up as an issue and prefix the title with "Proposal: "? Then we could discuss the implementation.
